tsl/src/remote/dist_copy.c

```diff
@@ -797,36 +1286,173 @@ send_copy_data(StringInfo row_data, const List *connections)
+				data_node_rows->row_indices = palloc(sizeof(int) * context->batch_row_count);
+				data_nodes = lappend(data_nodes, data_node_rows);
+			}
 
-	chunk = get_target_chunk(ht, point, &context->connection_state);
-	connections = get_connections_for_chunk(context, chunk->fd.id, chunk->data_nodes, GetUserId());
+			Assert(data_node_rows->server_oid == chunk_data_node->foreign_server_oid);
 
-	/* for remote copy, we don't use chunk insert states on the AN.
-	 * so we need to explicitly set the chunk as unordered when copies
-	 * are directed to previously compressed chunks
+			data_node_rows->row_indices[data_node_rows->rows_total] = row_in_batch;
+			data_node_rows->rows_total++;
+		}
+	}
+
+	/*
+	 * Flush the previous batch to avoid growing the outgoing buffers
+	 * indefinitely if some data node is not keeping up. It would be more
+	 * efficient to check for buffer growth and only flush then, but libpq
+	 * doesn't provide a way to know the outgoing buffer size. It also doesn't
+	 * provide any way to control the outgoing buffer size.
+	 * Don't do it if we have ended the COPY above to create new chunks.
+	 * The number 11 is an arbitrary prime, growing the output buffer to at
+	 * most 11ki rows sounds reasonable.
 	 */
-	if (ts_chunk_is_compressed(chunk) && (!ts_chunk_is_unordered(chunk)))
-		ts_chunk_set_unordered(chunk);
+	if (!did_end_copy && context->batch_ordinal % 11 == 0)
+	{
+		flush_active_connections(&context->connection_state);
+	}
+
+	/*
+	 * Actually send the data to the data nodes. We don't interleave the data
+	 * nodes here, because the batches are relatively small.
+	 */
+	StringInfoData copy_data = { .data = palloc(MAX_BATCH_BYTES), .maxlen = MAX_BATCH_BYTES };
+	ListCell *lc;
+	foreach (lc, data_nodes)
+	{
+		DataNodeRows *dn = lfirst(lc);
+		TSConnectionId required_id = remote_connection_id(dn->server_oid, GetUserId());
+		Assert(dn->connection == NULL);
+		dn->connection = get_copy_connection_to_data_node(context, required_id);
+		PGconn *pg_conn = remote_connection_get_pg_conn(dn->connection);
+
+		resetStringInfo(&copy_data);
+		for (int row = 0; row < dn->rows_total; row++)
+		{
+			StringInfo row_data = context->batch_row_data[dn->row_indices[row]];
+			appendBinaryStringInfo(&copy_data, row_data->data, row_data->len);
+		}
+
+		/*
+		 * Send the copy data to the remote server.
+		 * It can't really return 0 ("would block") until it runs out
+		 * of memory. It just grows the buffer and tries to flush in
+		 * pqPutMsgEnd().
+		 */
+		int res = PQputCopyData(pg_conn, copy_data.data, copy_data.len);
+
+		if (res == -1)
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_CONNECTION_EXCEPTION),
+					 errmsg("could not send COPY data"),
+					 errdetail("%s", PQerrorMessage(pg_conn))));
+		}
 
-	return send_copy_data(context->row_data, connections);
+		/*
+		 * We don't have to specially flush the data here, because the flush is
+		 * attempted after finishing each protocol message (pqPutMsgEnd()).
+		 */
+	}
+
+	return true;
```

Comment:
I guess this return value is pointless and we can remove it, instead returning void?